yaml2requirement.py

```diff
@@ -0,0 +1,260 @@
+        """
+        pattern = re.compile(r'(===|==|!=|~=|<=?|>=?|===?)\s*(.*)')
+        str_match = pattern.findall(spec_with_text)
+        specifier, text = str_match[0]
+        replacements = (('<', '>'),
+                        ('>', '<'),
+                        ('!', '='),
+                        ('~', '!'),
+                        ('==', '!='))
+        for old, new in replacements:
+            if old in specifier and specifier != '===':
+                new_spec = specifier.replace(old, new)
+                break
+            elif specifier == '===':
+                new_spec = '==='
+                break
+        return (new_spec, text, specifier)
+
+
+    def _yaml_to_requirement(self, yaml:list, exclude: bool = False) -> set:
+        """Converts YAML defined package into packaging.requirements Requirement
+        which can be directly used with pip
+        - when exclude is set to True it makes opposite logic for Requirement to be excluded"""
+        yaml_list:list = yaml
+
+        requirements_set: set[Requirement] = set()
+
+        if not yaml_list:
+            return requirements_set
+
+        for package in yaml_list:
+            requirement_str_list = [f"{package['package_name']}"]
+
+            if 'version' in package:
+                if not isinstance(package['version'], list):
+                    new_spec, ver, old_spec = self._change_specifier_logic(package['version'])
+                    requirement_str_list.append(
+                        f'{new_spec}{ver}' if exclude else f'{old_spec}{ver}'
+                        )
+
+                else:
+                    version_list = []
+                    for elem in package['version']:
+                        new_spec, ver, old_spec = self._change_specifier_logic(elem)
+                        if exclude:
+                            version_list.append(f'{new_spec}{ver}')
+                        else:
+                            version_list.append(f'{old_spec}{ver}')
+
+                    requirement_str_list.append(','.join(version_list))
+
+            if 'platform' in package or 'python' in package:
+                requirement_str_list.append('; ')
+
+            if 'platform' in package and 'version' not in package:
+                if not isinstance(package['platform'], list):
+                    requirement_str_list.append((
+                        f"sys_platform != '{package['platform']}'" if exclude
+                        else f"sys_platform == '{package['platform']}'"
+                        ))
+
+                else:
+                    platform_list = (
+                        [f"sys_platform != '{plf}'" if exclude
+                         else f"sys_platform == '{plf}'" for plf in package['platform']]
+                        )
+
+                    requirement_str_list.append(' or '.join(platform_list))
+
+            if ('platform' in package or 'python' in package) and 'version' in package and exclude:
```

Comment:
You do a lot of very similar checks over and over. The function is very hard to understand and follow the logic.

Package is a dictionary, right? `'platform' in package` then involves a hash computation with linear complexity then hash lookup. And this is repeated over and over. Why don't join together ifs instead of partially checking the same thing?

Also why do you test the easiest thing (Boolean) the last? If exclude is False then it the previous three lookups were all unnecessary complexity.